pgrx/src/fcall.rs

```diff
@@ -0,0 +1,218 @@
+
+    //
+    // do some validation to catch the cases we don't/can't directly call
+    //
+
+    if !matches!(pg_proc.prokind(), ProKind::Function) {
+        // It only makes sense to directly call regular functions.  Calling aggregate or window
+        // functions is nonsensical
+        return Err(FCallError::UnsupportedFunctionType);
+    } else if pg_proc.proargmodes().iter().any(|mode| *mode != ProArgMode::In) {
+        // Right now we only know how to support arguments with the IN mode.  Perhaps in the
+        // future we can support IN_OUT and TABLE return types
+        return Err(FCallError::UnsupportedArgumentModes);
+    } else if retoid == pg_sys::INTERNALOID
+        || pg_proc.proargtypes().iter().any(|oid| *oid == pg_sys::INTERNALOID)
+    {
+        // No idea what to do with the INTERNAL type.  Generally it's just a raw pointer but pgrx
+        // has no way to express that with `IntoDatum`.  And passing around raw pointers seem
+        // unsafe enough that if someone needs to do that, they probably have the ability to
+        // re-implement this function themselves.
+        return Err(FCallError::InternalTypeNotSupported);
+    } else if !T::is_compatible_with(retoid) {
+        // the requested Oid of `T` is not compatible with the actual function return type
+        return Err(FCallError::IncompatibleReturnType(T::type_oid(), retoid));
+    }
+
+    // we're likely going to be able to call the function, so convert our arguments into Datums
+    let arg_datums = args.iter().map(|a| a.as_datum()).collect::<Vec<_>>();
+
+    // if the function is STRICT and at least one of our argument values is `None` (ie, NULL)...
+    // we must return `None` now and not call the function.  Passing a NULL argument to a STRICT
+    // function will likely crash Postgres
+    let isstrict = pg_proc.proisstrict();
+    if isstrict && arg_datums.iter().any(|d| d.is_none()) {
+        return Ok(None);
+    }
+
+    //
+    // The following code is Postgres-version specific.  Right now, it's compatible with v12+
+    // v11 will need a different implementation.
+    //
+    // NB:  Which I don't want to do since it EOLs in November 2023
+    //
+
+    unsafe {
+        // initialize a stack-allocated `FmgrInfo` instance
+        let mut flinfo = pg_sys::FmgrInfo::default();
+        pg_sys::fmgr_info(func_oid, &mut flinfo);
+
+        // heap allocate a `FunctionCallInfoBaseData` properly sized so there's enough room
+        // for `args.len()` arguments
+        let fcinfo = pg_sys::palloc0(
+            std::mem::size_of::<pg_sys::FunctionCallInfoBaseData>()
+                + std::mem::size_of::<pg_sys::NullableDatum>() * args.len(),
+        ) as *mut pg_sys::FunctionCallInfoBaseData;
+
+        // initialize it
+        let fcinfo_ref = fcinfo.as_mut().unwrap();
+        fcinfo_ref.flinfo = &mut flinfo;
+        fcinfo_ref.fncollation = pg_sys::InvalidOid; // TODO:  We need to get a collation from somewhere?
```

Comment:
Is it ever unsafe to use the wrong one? If not, I think we should offer a version that uses InvalidOid (which is the default), and one that allows the user to provide it.

Reply:
> Is it ever unsafe to use the wrong one?

Using `InvalidOid` with a function that takes or returns a TEXT-type argument flat out fails with an error.  So we gotta provide some legitimate value somehow.  Jubilee had some ideas on Discord.